Userland/watch.cpp

```diff
@@ -0,0 +1,171 @@
+#include <stdio.h>
+#include <sys/time.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+static int opt_interval = 2;
+static bool flag_noheader = false;
+static bool flag_beep_on_fail = false;
+static int exit_code = 0;
+static pid_t child_pid = -1;
+
+void handle_signal(int signal);
+int run_command(const Vector<const char*>& command);
+int64_t usecs_till(const struct timeval& now, const struct timeval& targ);
+struct timeval get_current_time();
+String build_header_string(const Vector<const char*>&, const struct timeval&);
+
+int main(int argc, char** argv)
+{
+    signal(SIGINT, handle_signal);
+    pledge("stdio proc exec", nullptr);
+
+    Vector<const char*> command;
+    Core::ArgsParser args_parser;
+    args_parser.add_option(opt_interval, "Amount of time between updates", "interval", 'n', "seconds");
+    args_parser.add_option(flag_noheader, "Turn off the header describing the command and interval", "no-title", 't');
+    args_parser.add_option(flag_beep_on_fail, "Beep if the command has a non-zero exit code", "beep", 'b');
+    args_parser.add_positional_argument(command, "Command to run", "command");
+    args_parser.parse(argc, argv);
+
+    struct timeval interval;
+    if (opt_interval <= 0) {
+        interval = { 0, 100000 };
+    } else {
+        interval = { opt_interval, 0 };
+    }
+
+    auto header = build_header_string(command, interval);
+    command.append(nullptr);
+
+    auto now = get_current_time();
+    auto next_run_time = now;
+    auto ran_immediately = false;
+    while (true) {
+        int usecs_to_sleep = usecs_till(now, next_run_time);
+        ran_immediately = usecs_to_sleep < 0;
+        while (usecs_to_sleep > 0) {
+            usleep(usecs_to_sleep);
+            now = get_current_time();
+            usecs_to_sleep = usecs_till(now, next_run_time);
+        }
+        // Clear the screen, then reset the cursor position to the top left
+        printf("\033[H\033[2J");
+        // Print the header
+        if (!flag_noheader) {
+            printf("%s\n\n", header.characters());
+        }
+        fflush(stdout);
+        if (run_command(command) != 0) {
+            exit_code = 1;
+            if (flag_beep_on_fail) {
+                printf("\a");
+                fflush(stdout);
+            }
+        }
+        if (ran_immediately) {
+            // Calculate the next run time based on the current time
+            timeval_add(now, interval, next_run_time);
+        } else {
+            // Calculate the next run time based on the previous run time
+            timeval_add(next_run_time, interval, next_run_time);
+        }
+        now = get_current_time();
+    }
+}
+
+String build_header_string(const Vector<const char*>& command, const struct timeval& interval)
+{
+    StringBuilder builder;
+    builder.appendf("Every %d", interval.tv_sec);
+    builder.appendf(".%ds: \'", interval.tv_usec / 100000);
+    builder.join(' ', command);
+    builder.append("\'");
```

Comment:
Would perhaps be useful to add some color and/or bold

Reply:
I've added bolding around the command, not sure what color would look good here. Maybe blue for the rest?

Reply:
I think I'm just going to stick with bolding it.